util/pkg/vfs/swiftfs.go

```diff
@@ -0,0 +1,453 @@
+		}
+	})
+	if err == nil {
+		return os.ErrExist
+	} else if !os.IsNotExist(err) {
+		return err
+	}
+
+	err = p.createBucket()
+	if err != nil {
+		return err
+	}
+
+	return p.WriteFile(data)
+}
+
+func (p *SwiftPath) createBucket() error {
+	done, err := RetryWithBackoff(swiftWriteBackoff, func() (bool, error) {
+		_, err := swiftcontainer.Get(p.client, p.bucket).Extract()
+		if err == nil {
+			return true, nil
+		}
+		if isSwiftNotFound(err) {
+			createOpts := swiftcontainer.CreateOpts{}
+			_, err = swiftcontainer.Create(p.client, p.bucket, createOpts).Extract()
+			return err == nil, err
+		}
+		return false, err
+	})
+	if err != nil {
+		return err
+	} else if done {
+		return nil
+	} else {
+		// Shouldn't happen - we always return a non-nil error with false.
+		return wait.ErrWaitTimeout
+	}
+}
+
+// ReadFile implements Path::ReadFile.
+func (p *SwiftPath) ReadFile() ([]byte, error) {
+	var ret []byte
+	done, err := RetryWithBackoff(swiftReadBackoff, func() (bool, error) {
+		glog.V(4).Infof("Reading file %q", p)
+
+		opt := swiftobject.DownloadOpts{}
+		result := swiftobject.Download(p.client, p.bucket, p.key, opt)
+		data, err := (&result).ExtractContent()
+		if err == nil {
+			ret = data
+			return true, nil
+		} else if isSwiftNotFound(err) {
+			return true, os.ErrNotExist
+		} else {
+			return false, fmt.Errorf("error reading %s: %v", p, err)
+		}
+	})
+	if err != nil {
+		return nil, err
+	} else if done {
+		return ret, nil
+	} else {
+		return nil, wait.ErrWaitTimeout
+	}
+}
+
+func (p *SwiftPath) readPath(opt swiftobject.ListOpts) ([]Path, error) {
+	var ret []Path
+	done, err := RetryWithBackoff(swiftReadBackoff, func() (bool, error) {
+		var paths []Path
+		pager := swiftobject.List(p.client, p.bucket, opt)
+		err := pager.EachPage(func(page pagination.Page) (bool, error) {
+			objects, err1 := swiftobject.ExtractInfo(page)
+			if err1 != nil {
+				return false, err1
+			}
+			for _, o := range objects {
+				child := &SwiftPath{
+					client: p.client,
+					bucket: p.bucket,
+					key:    o.Name,
+					hash:   o.Hash,
+				}
+				paths = append(paths, child)
+			}
+
+			return true, nil
+		})
+		if err != nil {
+			if isSwiftNotFound(err) {
+				return true, os.ErrNotExist
+			}
+			return false, fmt.Errorf("error listing %s: %v", p, err)
+		}
+		glog.V(8).Infof("Listed files in %v: %v", p, paths)
+		ret = paths
+		return true, nil
+	})
+	if err != nil {
+		return nil, err
+	} else if done {
+		return ret, nil
+	} else {
+		return nil, wait.ErrWaitTimeout
+	}
+}
+
+// ReadDir implements Path::ReadDir.
+func (p *SwiftPath) ReadDir() ([]Path, error) {
+	prefix := p.key
+	if !strings.HasSuffix(prefix, "/") {
+		prefix += "/"
+	}
+	opt := swiftobject.ListOpts{
+		Full: true,
+		Path: prefix,
+	}
+	return p.readPath(opt)
+}
+
+// ReadTree implements Path::ReadTree.
+func (p *SwiftPath) ReadTree() ([]Path, error) {
+	prefix := p.key
+	if !strings.HasSuffix(prefix, "/") {
+		prefix += "/"
+	}
+	opt := swiftobject.ListOpts{
```

Comment:
So ReadTree is supposed to be recursive (`ls -R`), and ReadDir is supposed to be non-recursive (`ls`).   Does swift support that?  On S3 for example, we set `Delimiter` for ReadDir, but not for ReadTree.

Reply:
yes, Swift support it. But it should pass different option to the method of "swiftobject.List". In my test environment, I create several objects as following
bucket: kops
a/a.txt
a/b/b1.txt
a/b/b2.txt
a/b/c/c.txt
if key=a, ReadDir will return a/a.txt; ReadTree will return all of the 4 objects.
if key=a/b, ReadDir will return a/b/b1.txt and a/b/b2.txt; ReadTree will return a/b/b1.txt, a/b/b2.txt and a/b/c/c.txt
if key=a/b/c, both ReadDir and ReadTree will return a/b/c/c.txt

Reply:
That's absolutely correct.  But looking at the two functions (ReadDir & ReadTree) I don't see a difference - am I just missing it?

Reply:
Yes, there is a very small difference.  The variable of 'prefix' is assigned to the different member of 'CreatOpts' in these two method respectively.  In ReadDir, CreateOpts.Path = prefix; but in ReadTree, CreateOpts.Prefix= prefix. Sorry to cause you misunderstanding.

Reply:
Oops - I see it now!  Path vs Prefiix.  Thank you :-)